File: src/javascript/algorithms/graph/3.pacificAtlantic/pacificAtlantic.js
//DFS: TC:O(rows * cols) SC:O(rows * cols)

function pacificAtlantic(heights) {
    let rows = heights.length, cols = heights[0].length;
    let result = [];

    let pacific = Array(rows).fill().map((row) => Array(cols).fill(false));
    let atlantic = Array(rows).fill().map((row) => Array(cols).fill(false));

    // first row & last row visit
    for(let c=0; c< cols; c++) {
        dfs(heights, 0, c, pacific, Number.NEGATIVE_INFINITY);
        dfs(heights, rows-1, c, atlantic, Number.NEGATIVE_INFINITY);
    }

    // first col & last col visit
    for(let r=0; r< cols; r++) {
        dfs(heights, r, 0, pacific, Number.NEGATIVE_INFINITY);
        dfs(heights, r, cols-1, atlantic, Number.NEGATIVE_INFINITY);
    }

    for(let r=0; r<rows; r++) {
        for(let c=0; c<cols; c++) {
            if(pacific[r][c] && atlantic[r][c]) {
                result.push([r, c]);
            }
        }
    }
    return result;
}

function dfs(heights, row, col, visited, prevHeight) {
    if(row < 0 || row >= heights.length || col < 0 || col >= heights[0].length || heights[row][col] < prevHeight || visited[row][col]) return;
    visited[row][col] = true;

    //down, top, right & left
    dfs(heights, row+1, col, visited, heights[row][col]);
    dfs(heights, row-1, col, visited, heights[row][col]);
    dfs(heights, row, col+1, visited, heights[row][col]);
    dfs(heights, row, col-1, visited, heights[row][col]);
}

let heights1 = [[3,4,5,3,6], [4,3,4,5,5], [1,3,6,4,2], [8,9,4,1,6], [9,3,3,1,5]];
let heights2 = [[2],[2]];
console.log(pacificAtlantic(heights1));
console.log(pacificAtlantic(heights2));

File: src/java1/algorithms/graph/pacificAtlantic/PacificAtlantic.java
package java1.algorithms.graph.pacificAtlantic;
import java.util.*;

public class PacificAtlantic {
    //DFS: TC:O(rows * cols) SC:O(rows * cols)
    private static List<List<Integer>> pacificAtlantic(int[][] heights) {
        int rows = heights.length, cols = heights[0].length;
        List<List<Integer>> result = new ArrayList<>();

        boolean[][] pacific = new boolean[rows][cols];
        boolean[][] atlantic = new boolean[rows][cols];

        //first row and last row visit
        for(int c = 0; c < cols; c++) {
            dfs(heights, 0, c, pacific, Integer.MIN_VALUE);
            dfs(heights, rows-1, c, atlantic, Integer.MIN_VALUE);
        }

        //first column and last column vist
        for(int r = 0; r < rows; r++) {
            dfs(heights, r, 0, pacific, Integer.MIN_VALUE);
            dfs(heights, r, cols-1, atlantic, Integer.MIN_VALUE);
        }

        for(int row=0; row<rows; row++) {
            for(int col=0; col<cols; col++) {
                if(pacific[row][col] && atlantic[row][col]) {
                    result.add(Arrays.asList(row, col));
                }
            }
        }
        return result;
    }

    private static void dfs(int[][] heights, int row, int col, boolean[][] visited, int prevHeight) {
        if(row < 0 || row >= heights.length || col < 0 || col >= heights[0].length ||
           heights[row][col] < prevHeight || visited[row][col]) return;
        visited[row][col] = true;

        //down, top, right, left traversal
        dfs(heights, row+1, col, visited, heights[row][col]);
        dfs(heights, row-1, col, visited, heights[row][col]);
        dfs(heights, row, col+1, visited, heights[row][col]);
        dfs(heights, row, col-1, visited, heights[row][col]);
    }

    public static void main(String[] args) {
        int[][] heights1 = {{3,4,5,3,6}, {4,3,4,5,5}, {1,3,6,4,2}, {8,9,4,1,6}, {9,3,3,1,5}};
        int[][] heights2 = {{2},{2}};
        System.out.println(pacificAtlantic(heights1));
        System.out.println(pacificAtlantic(heights2));
    }
}
